### PR TITLE
Authenticated User plug

### PR DESCRIPTION
**Pull request by alejandrodevs**
## Addresses issue: [#REG-42](https://amcoit.atlassian.net/browse/REG-42)

This `AuthenticatedUser` plug validates the access token in the request against the OpenID Connect Provider (IdP).

### CODE REVIEW

File: lib/ueberauth/strategy/amco.ex [hunk 1]
```diff
@@ -202,13 +202,7 @@ defmodule Ueberauth.Strategy.Amco do
...
   defp fetch_user(conn, token) do
     conn = put_private(conn, :amco_token, token)
-    put_private(conn, :amco_user, %{
-      "sub" => "10",
-      "email" => "alejandrodevs@gmail.com",
-      "last_name" => "Gutierrez",
-      "first_name" => "Alejandro",
-      "phone_number" => "+523310971195"
-    })
+    put_private(conn, :amco_user, %{})
```
Comment (alejandrodevs) [inline lib/ueberauth/strategy/amco.ex:205]: I removed this dummy data that I was used for testing. This function should validate the id_token or consumes the `/userinfo` (it is not implemented yet) endpoint in the OIDC Provider.

File: lib/ueberauth/strategy/amco/user.ex [hunk 2]
```diff
@@ -0,0 +1,15 @@
+defmodule Ueberauth.Strategy.Amco.User do
```
Comment (alejandrodevs) [inline lib/ueberauth/strategy/amco/user.ex:1]: This struct will be assigned in `conn.assigns[:current_user]` in the AuthenticatedUser plug.

File: lib/ueberauth/strategy/amco/api/access_token.ex [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+  alias Ueberauth.Strategy.Amco.OAuth
+
+  def authorize(token) do
+    OAuth.authorize_access_token(token)
+    |> process_response()
+  end
+
+  defp process_response({:ok, %Response{status_code: 200} = response}) do
+    case response.body do
+      %{"active" => true} = data -> {:ok, data}
```
Comment (javierg) [inline lib/ueberauth/strategy/amco/api/access_token.ex]: This is the same as `%{"active" => true} -> {:ok, response.body}`

IMO data is a very broad  term, so if you want to express the kind of response you are setting, maybe look for a different word, or just leave the response.body  as the returned data?

File: lib/ueberauth/strategy/amco/plugs/authenticated_user.ex [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+  If the access token is valid, the current user will be assigned to the
+  connection and the request will continue as normal. In the other hand,
+  it will call the access_token_error function in the response handler.
+  """
+
+  def init(opts), do: opts
+
+  def call(%Conn{} = conn, opts) do
+    format = Keyword.get(opts, :format, :html)
+    handler = Keyword.fetch!(opts, :response_handler)
```
Comment (javierg) [inline lib/ueberauth/strategy/amco/plugs/authenticated_user.ex]: in elixir is uncommon to let a function throw an exception, you usually return an error tuple when something is not set.

File: lib/ueberauth/strategy/amco/plugs/authenticated_user.ex [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+    with {:ok, token} <- get_access_token(conn, format),
+         {:ok, response} <- validate_access_token(token) do
+
+      if Keyword.has_key?(handler.__info__(:functions), :access_token_success) do
+        handler.access_token_success(conn, response)
+      end
+
+      conn
```
Comment (javierg) [inline lib/ueberauth/strategy/amco/plugs/authenticated_user.ex:30]: can the logic inside the with block be in a different function, I find confusing reading if end and then else, also the logic inside the if condition if kind of complex, you are retrieving the :functions key from the handler info, and then checking if it has a the access token success key, IMO that alone deserves its own function

File: lib/ueberauth/strategy/amco/plugs/authenticated_user.ex [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    case Conn.get_req_header(conn, "authorization") do
+      ["Bearer " <> access_token] -> {:ok, access_token}
+      _ -> {:error, :access_token_required}
+    end
+  end
+
+  defp get_access_token(%Conn{} = conn, :html) do
+    case Conn.get_session(conn, :current_user) do
+      nil -> {:error, :access_token_required}
+      user -> {:ok, user.credentials.token}
```
Comment (javierg) [inline lib/ueberauth/strategy/amco/plugs/authenticated_user.ex]: the user is expected with certain structure? can you match it here? and return error if it is not as expected?

File: lib/ueberauth/strategy/amco/api/access_token.ex (hunk 3 above)
Comment (alejandrodevs) [reply lib/ueberauth/strategy/amco/api/access_token.ex]: Fixed, thanks.

File: lib/ueberauth/strategy/amco/plugs/authenticated_user.ex (hunk 6 above)
Comment (alejandrodevs) [reply lib/ueberauth/strategy/amco/plugs/authenticated_user.ex]: Fixed. Thanks.

File: lib/ueberauth/strategy/amco/plugs/authenticated_user.ex (hunk 4 above)
Comment (alejandrodevs) [reply lib/ueberauth/strategy/amco/plugs/authenticated_user.ex]: I refactored it. Thanks.

File: lib/ueberauth/strategy/amco/plugs/authenticated_user.ex (hunk 5 above)
Comment (alejandrodevs) [reply lib/ueberauth/strategy/amco/plugs/authenticated_user.ex:30]: I removed the success callback because it is not required by the moment. Thanks.

**Review by javierg — APPROVED** [review]
:shipit: 

nicely done!

**Pull request merged**